Update all dependencies (except @types/node and jest-chance)

liquidnya opened the pull request:
### Checklist

- [x] Have you followed the guidelines in our Contributing document?
- [x] Have you run `eslint` on the code and resolved any errors?
- [x] Have you run `npm test` and all tests are passing?
- [ ] Have you added new tests for any new functions created?
- [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you updated CHANGELOG.md to add your changed under the `[Unreleased]` heading?

### Description

This PR is based on #84 (because of a lot of changes in `package.json` in that branch).

Updates all dependencies to the newest version!
Except:
- Downgrades `@types/node` to `^18.6.0` (the minimum version of node for this project).
- Does not upgrade jest-chance, because of its minimum node version `>=19.0.0` requirement.

Also moved `esbuild` to the dev dependencies.
Changed `moduleResolution` to `node16` again (in `tsconfig.json`), because the new version of `@tsconfig/node18/tsconfig.json` now declares `"module": "Node16"`.

### Benefits

Upgraded versions, and also more consistent setup.

### Potential drawbacks

This could break something.

demize commented:
Can you rebase on develop so that the changes are up to date? I should be able to review and approve after.

liquidnya commented:
I also changed the `@types/node` version to be `~18.6.0` instead of `^18.6.0`, this means that currently `@types/node` version `18.6.5` is used instead of `18.16.12`, which I think is the better choice if we allow node version `18.6.0` as the minimum version.
I also updated the dependencies in `package-lock.json` with `npm update`.

liquidnya commented:
Documentation of how to update versions (for future reference):
1. Use `npm outdated` to see a list of outdated versions
2. Update to the latest version except if those have breaking changes or require a node version higher than our minimum node version
3. Use `npm update` to update all versions in `package-lock.json`

If packages have breaking changes:
- You can either make changes to the code to fix those breaking changes.
- Delay updating that dependency.
- Create a new issue (<https://github.com/fluid-queue/fluid-queue/issues>) and the update will be made in a different PR.

If packages require a higher version than our minimum version:
- To test this one can use our minimum version for example exactly version `18.6.0` and then run `npm install`.
   If an error like this is printed then we either need to update our minimum version or delay the update of that dependency:
   ```
   npm WARN EBADENGINE Unsupported engine {
  npm WARN EBADENGINE   package: 'jest-chance@0.2.5',
  npm WARN EBADENGINE   required: { node: '>=19.0.0' },
  npm WARN EBADENGINE   current: { node: 'v18.6.0', npm: '8.13.2' }
  npm WARN EBADENGINE }
   ```

For future reference how to update `jest-chance`:
If tests do not work anymore (`npm test`), then one needs to change the value of `"globalSetup"` in `tests/jest.config.json` from `"jest-chance"` to `"./node_modules/jest-chance/dist/esm/index.js"`.
This is because jest includes the commonjs version instead of the esm version and then it crashes, because a module was expected.
You can also revert the following commit: <https://github.com/fluid-queue/fluid-queue/pull/86/commits/b9f8596ed3f6571db82368ff14ae9a197c1ed15d>.

liquidnya commented:
How to update to a new minimum node version:
- Make sure to update `"engines"`/`"node"` in `package.json`.
- Update `@types/node` to the minimum version prefixed with `~`.
- Change the `@tsconfig/node18` dependency to the correct major version in its name and update the `"extends"` field of `tsconfig.json` and make sure to adjust `"compilerOptions"` too if needed.
- Change the target version in `build.ts` to be the minimum node version.

demize reviewed (approved):
LGTM. We should probably document your instructions there somewhere better; maybe use the Github Wiki as somewhere to keep details like that?
